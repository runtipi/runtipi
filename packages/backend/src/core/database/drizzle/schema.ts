import { relations } from 'drizzle-orm';
import { boolean, customType, foreignKey, integer, pgEnum, pgTable, serial, text, timestamp, unique, varchar } from 'drizzle-orm/pg-core';

export const appStatusEnum = pgEnum('app_status_enum', [
  'running',
  'stopped',
  'installing',
  'uninstalling',
  'stopping',
  'starting',
  'missing',
  'updating',
  'resetting',
  'restarting',
  'backing_up',
  'restoring',
]);
export const updateStatusEnum = pgEnum('update_status_enum', ['FAILED', 'SUCCESS']);

export const link = pgTable(
  'link',
  {
    id: serial().primaryKey().notNull(),
    title: varchar({ length: 20 }).notNull(),
    url: varchar().notNull(),
    iconUrl: varchar('icon_url'),
    createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    userId: integer('user_id').notNull(),
    description: varchar({ length: 50 }),
  },
  (table) => {
    return {
      fkLinkUserId: foreignKey({
        columns: [table.userId],
        foreignColumns: [user.id],
        name: 'FK_link_user_id',
      }).onDelete('cascade'),
    };
  },
);

const appConfig = customType<{ data: Record<string, unknown>; driverData: string }>({
  dataType() {
    return 'jsonb';
  },
  toDriver(value: Record<string, unknown>): string {
    return JSON.stringify(value);
  },
});

export const app = pgTable(
  'app',
  {
    id: varchar().primaryKey().notNull(),
    status: appStatusEnum().default('stopped').notNull(),
    lastOpened: timestamp({ withTimezone: true, mode: 'string' }).defaultNow(),
    numOpened: integer().default(0).notNull(),
    config: appConfig('config').notNull(),
    createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    version: integer().default(1).notNull(),
    exposed: boolean().default(false).notNull(),
    domain: varchar(),
    isVisibleOnGuestDashboard: boolean('is_visible_on_guest_dashboard').default(false).notNull(),
    openPort: boolean('open_port').default(true).notNull(),
    exposedLocal: boolean('exposed_local').default(true).notNull(),
    appStoreId: integer('app_store_id').references(() => appStore.id),
  },
  (table) => {
    return {
      uq9478629Fc093D229Df09E560Aea: unique('UQ_9478629fc093d229df09e560aea').on(table.id),
    };
  },
);

export const appRelations = relations(app, ({ one }) => ({
  appStore: one(appStore, {
    fields: [app.appStoreId],
    references: [appStore.id],
  }),
}));

export const user = pgTable(
  'user',
  {
    id: serial().primaryKey().notNull(),
    username: varchar().notNull(),
    password: varchar().notNull(),
    createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
    operator: boolean().default(false).notNull(),
    totpSecret: text('totp_secret'),
    totpEnabled: boolean('totp_enabled').default(false).notNull(),
    salt: text(),
    locale: varchar().default('en').notNull(),
    hasSeenWelcome: boolean('has_seen_welcome').default(false).notNull(),
  },
  (table) => {
    return {
      uq78A916Df40E02A9Deb1C4B75Edb: unique('UQ_78a916df40e02a9deb1c4b75edb').on(table.username),
    };
  },
);

export const appStore = pgTable('app_store', {
  id: serial().primaryKey().notNull(),
  hash: varchar().notNull().unique(),
  name: varchar().notNull(),
  enabled: boolean().default(true).notNull(),
  url: varchar().notNull(),
  branch: varchar().default('main').notNull(),
  createdAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
  updatedAt: timestamp({ mode: 'string' }).defaultNow().notNull(),
  deleted: boolean().default(false).notNull(),
});
