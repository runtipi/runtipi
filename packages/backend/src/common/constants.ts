export const APP_DIR = '/app';
export const DATA_DIR = '/data';
export const APP_DATA_DIR = '/app-data';

export const SESSION_COOKIE_NAME = 'tipi.sid';
export const SESSION_COOKIE_MAX_AGE = 1000 * 60 * 60 * 24;

export const ARCHITECTURES = ['arm64', 'amd64'] as const;
export type Architecture = (typeof ARCHITECTURES)[number];

export const LATEST_RELEASE_URL = 'https://api.github.com/repos/runtipi/runtipi/releases/latest';
