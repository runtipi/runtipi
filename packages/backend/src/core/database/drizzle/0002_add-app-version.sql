-- Create version field if it doesn't exist
ALTER TABLE "app"
    ADD COLUMN IF NOT EXISTS "version" integer DEFAULT '0';

-- Set version field to 1 for all apps that have no version
UPDATE
    "app"
SET
    "version" = '1'
WHERE
    "version" IS NULL
    OR "version" = '0';

-- Set version field to not null
ALTER TABLE "app"
    ALTER COLUMN "version" SET NOT NULL;

DO $$
BEGIN
    IF NOT EXISTS (
        SELECT
            *
        FROM
            information_schema.table_constraints
        WHERE
            constraint_name = 'UQ_9478629fc093d229df09e560aea'
            AND table_name = 'app') THEN
    ALTER TABLE "app"
        ADD CONSTRAINT "UQ_9478629fc093d229df09e560aea" UNIQUE ("id");
END IF;
END
$$;


