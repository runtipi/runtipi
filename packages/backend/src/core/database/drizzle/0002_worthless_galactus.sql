DROP TABLE IF EXISTS "migrations" CASCADE;--> statement-breakpoint
DROP TABLE IF EXISTS "update" CASCADE;
