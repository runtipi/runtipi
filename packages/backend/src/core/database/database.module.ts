import { Global, Module } from '@nestjs/common';
import { DatabaseService } from './database.service';

export const DATABASE = 'DATABASE_INSTANCE';
export type Database = DatabaseService['db'];

@Global()
@Module({
  imports: [],
  controllers: [],
  providers: [
    DatabaseService,
    {
      provide: DATABASE,
      useFactory: (databaseService: DatabaseService): Database => databaseService.db,
      inject: [DatabaseService],
    },
  ],
  exports: [DatabaseService, DATABASE],
})
export class DatabaseModule {}
