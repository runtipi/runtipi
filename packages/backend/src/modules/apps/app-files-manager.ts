import path from 'node:path';
import { extractAppId } from '@/common/helpers/app-helpers';
import { execAsync } from '@/common/helpers/exec-helpers';
import { ConfigurationService } from '@/core/config/configuration.service';
import { FilesystemService } from '@/core/filesystem/filesystem.service';
import { LoggerService } from '@/core/logger/logger.service';
import { Injectable } from '@nestjs/common';
import { appInfoSchema } from '../marketplace/dto/marketplace.dto';

@Injectable()
export class AppFilesManager {
  constructor(
    private readonly configuration: ConfigurationService,
    private readonly filesystem: FilesystemService,
    private readonly logger: LoggerService,
  ) {}

  private getInstalledAppsFolder() {
    const { directories } = this.configuration.getConfig();

    return path.join(directories.dataDir, 'apps');
  }

  public getAppPaths(namespacedAppId: string) {
    const { directories } = this.configuration.getConfig();

    const { storeId, appId } = extractAppId(namespacedAppId);

    return {
      appDataDir: path.join(directories.appDataDir, storeId, appId),
      appInstalledDir: path.join(this.getInstalledAppsFolder(), storeId, appId),
    };
  }

  /**
   * Get the app info from the installed apps apps
   * @param id - The app id
   */
  public async getInstalledAppInfo(id: string) {
    try {
      const { appInstalledDir } = this.getAppPaths(id);

      if (await this.filesystem.pathExists(path.join(appInstalledDir, 'config.json'))) {
        const configFile = await this.filesystem.readTextFile(path.join(appInstalledDir, 'config.json'));
        const parsedConfig = appInfoSchema.safeParse(JSON.parse(configFile ?? ''));

        if (!parsedConfig.success) {
          this.logger.debug(`App ${id} config error:`);
          this.logger.debug(parsedConfig.error);
        }

        if (parsedConfig.success && parsedConfig.data.available) {
          const description = (await this.filesystem.readTextFile(path.join(appInstalledDir, 'metadata', 'description.md'))) ?? '';

          return { ...parsedConfig.data, id, description };
        }
      }
    } catch (error) {
      return null;
    }
  }

  /**
   * Get the docker-compose.json file content from the installed app
   * @param id - The app id
   * @returns The content of docker-compose.yml as a string, or null if not found
   */
  public async getDockerComposeYaml(id: string) {
    const arch = this.configuration.get('architecture');
    const { appInstalledDir } = this.getAppPaths(id);
    let dockerComposePath = path.join(appInstalledDir, 'docker-compose.yml');

    if (arch === 'arm64' && (await this.filesystem.pathExists(path.join(appInstalledDir, 'docker-compose.arm64.yml')))) {
      dockerComposePath = path.join(appInstalledDir, 'docker-compose.arm64.yml');
    }

    let content = null;
    try {
      if (await this.filesystem.pathExists(dockerComposePath)) {
        content = await this.filesystem.readTextFile(dockerComposePath);
      }
    } catch (error) {
      this.logger.error(`Error getting docker-compose.yml for installed app ${id}: ${error}`);
    }

    return { path: dockerComposePath, content };
  }

  /**
   * Get the docker-compose.json file content from the installed app
   * @param id - The app id
   * @returns The content of docker-compose.json as a string, or null if not found
   */
  public async getDockerComposeJson(id: string) {
    const { appInstalledDir } = this.getAppPaths(id);
    const dockerComposePath = path.join(appInstalledDir, 'docker-compose.json');

    let content = null;
    try {
      if (await this.filesystem.pathExists(dockerComposePath)) {
        content = await this.filesystem.readJsonFile(dockerComposePath);
      }
    } catch (error) {
      this.logger.error(`Error getting docker-compose.json for installed app ${id}: ${error}`);
    }

    return { path: dockerComposePath, content };
  }

  /**
   * Write the docker-compose.yml file to the installed app folder
   * @param appId - The app id
   * @param composeFile - The content of the docker-compose.yml file
   */
  public async writeDockerComposeYml(appId: string, composeFile: string) {
    const { appInstalledDir } = this.getAppPaths(appId);
    const dockerComposePath = path.join(appInstalledDir, 'docker-compose.yml');

    await this.filesystem.writeTextFile(dockerComposePath, composeFile);
  }

  public async deleteAppFolder(appId: string) {
    const { appInstalledDir } = this.getAppPaths(appId);
    await this.filesystem.removeDirectory(appInstalledDir);
  }

  public async deleteAppDataDir(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);
    await this.filesystem.removeDirectory(appDataDir);
  }

  public async createAppDataDir(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);
    await this.filesystem.createDirectory(appDataDir);
  }

  /**
   * Set the permissions for the app data directory
   * @param appId - The app id
   */
  public async setAppDataDirPermissions(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);

    await execAsync(`chmod -Rf a+rwx ${appDataDir}`).catch(() => {
      this.logger.error(`Error setting permissions for app ${appId}`);
    });
  }

  public async getAppEnv(appId: string) {
    const { appDataDir } = this.getAppPaths(appId);

    const envPath = path.join(appDataDir, 'app.env');

    let env = '';
    if (await this.filesystem.pathExists(envPath)) {
      env = (await this.filesystem.readTextFile(envPath)) ?? '';
    }

    return { path: envPath, content: env };
  }

  public async writeAppEnv(appId: string, env: string) {
    const { appDataDir } = this.getAppPaths(appId);

    const envPath = path.join(appDataDir, 'app.env');

    await this.filesystem.writeTextFile(envPath, env);
  }

  /**
   * Get the user env file content
   * @param namespacedAppId - The app id
   */
  public async getUserEnv(namespacedAppId: string) {
    const { directories } = this.configuration.getConfig();

    const { storeId, appId } = extractAppId(namespacedAppId);

    const userEnvFile = path.join(directories.dataDir, 'user-config', storeId, appId, 'app.env');
    let content = null;

    if (await this.filesystem.pathExists(userEnvFile)) {
      content = await this.filesystem.readTextFile(userEnvFile);
    }

    return { path: userEnvFile, content };
  }

  /**
   * Get the user compose file content
   * @param namespacedAppId - The app id
   */
  public async getUserComposeFile(namespacedAppId: string) {
    const { directories } = this.configuration.getConfig();

    const { storeId, appId } = extractAppId(namespacedAppId);

    const userComposeFile = path.join(directories.dataDir, 'user-config', storeId, appId, 'docker-compose.yml');
    let content = null;

    if (await this.filesystem.pathExists(userComposeFile)) {
      content = await this.filesystem.readTextFile(userComposeFile);
    }

    return { path: userComposeFile, content };
  }
}
