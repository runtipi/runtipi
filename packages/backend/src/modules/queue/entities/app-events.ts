import { appFormSchema } from '@/modules/app-lifecycle/dto/app-lifecycle.dto';
import type { AppUrn } from '@/types/app/app.types';
import { Injectable } from '@nestjs/common';
import { type ZodStringDef, z } from 'zod';
import { Queue } from '../queue.entity';

const commonAppCommandSchema = z.object({
  command: z.union([
    z.literal('start'),
    z.literal('stop'),
    z.literal('install'),
    z.literal('uninstall'),
    z.literal('reset'),
    z.literal('restart'),
    z.literal('generate_env'),
    z.literal('backup'),
  ]),
  appUrn: z.string().refine((v) => v.split(':').length === 2) as unknown as z.ZodType<AppUrn, ZodStringDef>,
  skipEnv: z.boolean().optional().default(false),
  form: appFormSchema,
});

const restoreAppCommandSchema = z.object({
  command: z.literal('restore'),
  appUrn: z.string().refine((v) => v.split(':').length === 2) as unknown as z.ZodType<AppUrn, ZodStringDef>,
  filename: z.string(),
  form: appFormSchema,
});

const updateAppCommandSchema = z.object({
  command: z.literal('update'),
  appUrn: z.string().refine((v) => v.split(':').length === 2) as unknown as z.ZodType<AppUrn, ZodStringDef>,
  form: appFormSchema,
  performBackup: z.boolean().optional().default(true),
});

export const appEventSchema = commonAppCommandSchema.or(restoreAppCommandSchema).or(updateAppCommandSchema);

export const appEventResultSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type AppEventFormInput = z.input<typeof commonAppCommandSchema>['form'];
export type AppEventForm = z.output<typeof commonAppCommandSchema>['form'];

@Injectable()
export class AppEventsQueue extends Queue<typeof appEventSchema, typeof appEventResultSchema> {}
