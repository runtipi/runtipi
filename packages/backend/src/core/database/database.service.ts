import path from 'node:path';
import { Injectable } from '@nestjs/common';
import { type NodePgDatabase, drizzle } from 'drizzle-orm/node-postgres';
import { migrate } from 'drizzle-orm/node-postgres/migrator';
import { ConfigurationService } from '../config/configuration.service';
import { LoggerService } from '../logger/logger.service';
import * as schema from './drizzle/schema';

@Injectable()
export class DatabaseService {
  public db: NodePgDatabase<typeof schema>;

  constructor(
    private configurationService: ConfigurationService,
    private logger: LoggerService,
  ) {
    const { username, port, database, host, password } = this.configurationService.get('database');
    const connectionString = `postgresql://${username}:${password}@${host}:${port}/${database}?connect_timeout=300`;

    this.db = drizzle(connectionString, { schema });
  }

  private getMigrationsPath(): string {
    const { appDir } = this.configurationService.get('directories');
    return path.join(appDir, 'assets', 'migrations');
  }

  migrate = async () => {
    try {
      this.logger.debug('Starting database migration...');
      await migrate(this.db, { migrationsFolder: this.getMigrationsPath() });
      this.logger.debug('Database migration complete.');
    } catch (error) {
      this.logger.error('Error migrating database:', error);
      throw error;
    }
  };
}
