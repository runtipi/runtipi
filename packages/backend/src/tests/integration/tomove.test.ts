import fs from 'node:fs';
import { APP_DATA_DIR, APP_DIR, DATA_DIR } from '@/common/constants';
import { ConfigurationService } from '@/core/config/configuration.service';
import { DatabaseService } from '@/core/database/database.service';
import { appStore } from '@/core/database/drizzle/schema';
import { FilesystemService } from '@/core/filesystem/filesystem.service';
import { AppLifecycleCommandFactory } from '@/modules/app-lifecycle/app-lifecycle-command.factory';
import { AppLifecycleService } from '@/modules/app-lifecycle/app-lifecycle.service';
import { AppStoreRepository } from '@/modules/app-stores/app-store.repository';
import { AppStoreService } from '@/modules/app-stores/app-store.service';
import { AppFilesManager } from '@/modules/apps/app-files-manager';
import { AppsRepository } from '@/modules/apps/apps.repository';
import { MarketplaceService } from '@/modules/marketplace/marketplace.service';
import { AppEventsQueue, appEventResultSchema, appEventSchema } from '@/modules/queue/entities/app-events';
import { QueueFactory } from '@/modules/queue/queue.factory';
import { Test } from '@nestjs/testing';
import { fromPartial } from '@total-typescript/shoehorn';
import { beforeAll, beforeEach, describe, expect, it } from 'vitest';
import { mock } from 'vitest-mock-extended';
import waitFor from 'wait-for-expect';
import type { FsMock } from '../__mocks__/fs';
import { createAppInStore } from '../utils/create-app-in-store';
import { createTestDatabase } from '../utils/create-test-database';

let db: Awaited<ReturnType<typeof createTestDatabase>>;

describe('App lifecycle', () => {
  let appLifecycleService: AppLifecycleService;
  let marketplaceService: MarketplaceService;
  let appsRepository: AppsRepository;
  let configurationService = mock<ConfigurationService>();
  let databaseService = mock<DatabaseService>();

  beforeAll(async () => {
    db = await createTestDatabase('integrationtest');
  });

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      providers: [
        AppLifecycleService,
        MarketplaceService,
        AppStoreService,
        AppStoreRepository,
        FilesystemService,
        QueueFactory,
        AppLifecycleCommandFactory,
        AppFilesManager,
        {
          provide: AppEventsQueue,
          useFactory: (queueFactory: QueueFactory) =>
            queueFactory.createQueue({
              queueName: 'app-events-queue',
              workers: 1,
              eventSchema: appEventSchema,
              resultSchema: appEventResultSchema,
            }),
          inject: [QueueFactory],
        },
      ],
    })
      .useMocker(mock)
      .compile();

    appLifecycleService = moduleRef.get(AppLifecycleService);
    configurationService = moduleRef.get(ConfigurationService);
    databaseService = moduleRef.get(DatabaseService);
    marketplaceService = moduleRef.get(MarketplaceService);
    appsRepository = moduleRef.get(AppsRepository);

    databaseService.db = db;
    configurationService.getConfig.mockReturnValue(
      fromPartial({ demoMode: false, directories: { dataDir: DATA_DIR, appDir: APP_DIR, appDataDir: APP_DATA_DIR } }),
    );

    db.insert(appStore).values({ id: 1, url: 'https://appstore.example.com', hash: 'test', name: 'test', enabled: true }).execute();

    await marketplaceService.initialize();
  });

  describe('install app', () => {
    it('should copy the correct files and folders from app store', async () => {
      // arrange
      const appInfo = await createAppInStore(1, { id: 'test' });
      const appId = `1_${appInfo.id}`;

      // act
      await appLifecycleService.installApp({ appId, form: {} });
      await new Promise((resolve) => setTimeout(resolve, 2000));

      waitFor(async () => {
        const app = await appsRepository.getApp(appId);
        expect(app).toBeDefined();
      });

      // assert
      expect((fs as unknown as FsMock).tree()).toMatchSnapshot();
    });
  });
});
