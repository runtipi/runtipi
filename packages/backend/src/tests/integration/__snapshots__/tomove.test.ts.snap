// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App lifecycle > install app > should copy the correct files and folders from app store 1`] = `
"/
├─ app-data/
│  └─ 1/
│     └─ test/
│        └─ data/
└─ data/
   ├─ apps/
   │  └─ 1/
   │     └─ test/
   │        ├─ config.json
   │        ├─ data/
   │        └─ docker-compose.json
   ├─ backups/
   ├─ repos/
   │  └─ 1/
   │     └─ apps/
   │        └─ test/
   │           ├─ config.json
   │           ├─ data/
   │           └─ docker-compose.json
   └─ state/
      └─ seed"
`;
